### DISCUSSION TITLE
Speed for motors w rotation sensors capped?

### DISCUSSION THREAD

**Discussion by andybellgbc** [pybricks/support#1620] [Q&A] [answered]
I'm working on a primer for getting started with GBC. It's one of the most common questions asked at displays.

The LEGO motor and control availability is not straight forward... And with GBC looking for a rate (80 to 60 RPM ) knowing motor speed and controlling that is an important element.

To get a handle on what's out there I have been running some tests.

It appears with the technic hub and the boost motor the speed is capped at 990 deg/sec (165 rpm). With the Control Lg motor the speed is capped at 1500 deg/sec (250 rpm) regardless of number entered.

Using the LEGO app and creating a simple custom program I am able to run those motors at 260 and 320 rpm which is the same speed as the technic switched hub.

Is there a reason for these caps? Can they be extended? see note

It would sure be helpful if they could.

fyi the less sophisticated PUP DC motor is much closer in these tests 370 rpm in Pybricks, 380 in the technic hub/app and 390 in the technic switch.

Note I'm pretty sure this is a software implementation limit because regardless of power supply - rechargeable batteries, alkaline batteries or 9v wall adapter the output from Pybricks is the same (including the city hub), but all other strategies LEGO app or switched hub the rotation speed varies some.

![CtrlLgTestedv1 0](https://github.com/pybricks/support/assets/168368037/be5193fe-162d-4c34-8b78-751fd2f115aa)
![BoostTestedv1 0](https://github.com/pybricks/support/assets/168368037/e5f7a507-7ab2-45bb-b713-459f88516ff2)

**Comment by dlech** [comment]
What functions are you using for your tests? `motor.dc(100)`?

**Comment by laurensvalk** [comment, marked as answer]
That's some detailed analysis! Thanks for sharing.

> Using the LEGO app and creating a simple custom program

Which blocks did you use in the LEGO app? To compare apples to apples, you'd want to compare "100% power" to `motor.dc(100)`.

I don't expect much difference between the apps. This is the equivalent of applying the full batteries to the motor. I think this is what you found when you said:

> fyi the less sophisticated PUP DC motor is much closer in these tests 370 rpm in Pybricks, 380 in the technic hub/app and 390 in the technic switch.

------------

And then compare the "100% regulated speed controlled block" to the `motor.run(...)` methods. These do a have a software limit, so that it has some breathing room of still regulating the speed with various battery types. It's possible that the limit varies a bit between apps.

But you can [increase these limits](https://docs.pybricks.com/en/latest/pupdevices/motor.html#pybricks.pupdevices.Motor.Motor.control.limits) if you really want to. Increasing the acceleration can be useful, but increasing the speed cap usually isn't useful as explained above. It creates the false impression of being able to reach a speed you can't really reach, which means you'll get less accurate speed/position control.

It's possible that a few motors could use improved defaults. If you see major differences for certain motor types, we can look at that.

**Comment by andybellgbc** [comment]
Thank you for the helpful answers.

I started with the sample code from Huw's article

https://brickset.com/article/97497/externally-powering-the-technic-hub

which uses   -  motor.run(speed)

Which I have used for the Boost and Control Lg motors by increasing the speed and measuring with an inexpensive motor shaft tach, I could see where the speed/rotation seem to be capped at 990 and 1500

through reading the documentation I used for the PuP motor tests - (Make the motor go clockwise (forward) at 100% duty cycle ("100% power"))

pupmotor = DCMotor(Port.B)
pupdc = 100
pupmotor.dc(pupdc)

I did not try the DCmotor call on the Boost and Control Lg motors - I assumed they would not work and apparently missed it in the documentation.  The LEGO app appears to be using the % duty cycle type function for all motor types. 

I also see the link to setting the peak speed, which I missed.

I agree that setting the limit higher than what is possible is not helpful and may cause confusion. However the same is true for setting it significantly lower... (I would think the Boost M at 200 rpm would work for all power types. 165 seems unnecessarily low)  Which of course begs the question: "Which power source is the baseline?" I can understand rechargeable batteries as they tend to be lower than true 9v or fresh alkaline, and are likely for someone looking to use the hub enough to install pybricks.

Having missed the part about being able to set the maximum speed, is my oversight. Stating the software default maximum, perhaps in the documentation about setting it higher, would save people from trying to figure out the default value.
 
As for my purpose - GBC - being able to set a stable/consistent speed is much more beneficial that setting a % of power with the DCMotor   function, even with a stable power supply vs batteries.

I will investigate setting the maximum and see how that works.

Thank you for the helpful answers.

**Reply by laurensvalk** [reply]
It's quite possible that the cap on the external Boost motor is a bit too strict. I haven't used that motor extensively, so it could be an oversight as I was adding all the settings all at once. I'll open an issue to look into it and bump the limit to a more reasonable value.

**Reply by andybellgbc** [reply]
Another related question. It's clear the software is aware of which motor is plugged in as the caps are different by type.  Is there a way to get that data and use it?  For example if BoostExt then speed=990 if ControlLg = 1440 else 1200 or similar.

**Reply by laurensvalk** [reply]
To get the max speed for any motor, just do:

```python

my_motor = Motor(Port.A)

speed, acceleration, torque = my_motor.control.limits()

print(speed)

# Run at max rated speed
my_motor.run(speed)


```

To get the device ID, try this:

```python
from pybricks.iodevices import PUPDevice
from pybricks.parameters import Port

device = PUPDevice(Port.A)

# Get the device id
id = device.info()["id"]
print(id)
```
